File: services/risk-assessment-service/sdks/nodejs/src/exceptions.js
/**
 * KYB Platform Risk Assessment Service Node.js SDK Exceptions
 * 
 * This module defines custom exceptions for the KYB Platform SDK.
 */

/**
 * Base exception for all KYB SDK errors.
 */
class KYBError extends Error {
    constructor(message, errorData = {}) {
        super(message);
        this.name = this.constructor.name;
        this.message = message;
        this.errorData = errorData;
        
        // Maintains proper stack trace for where our error was thrown (only available on V8)
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, this.constructor);
        }
    }

    /**
     * Get the request ID from the error data.
     * @returns {string|null} The request ID
     */
    getRequestId() {
        return this.errorData.request_id || null;
    }

    /**
     * Get the timestamp from the error data.
     * @returns {string|null} The timestamp
     */
    getTimestamp() {
        return this.errorData.timestamp || null;
    }

    /**
     * Get the request path from the error data.
     * @returns {string|null} The request path
     */
    getPath() {
        return this.errorData.path || null;
    }

    /**
     * Get the request method from the error data.
     * @returns {string|null} The request method
     */
    getMethod() {
        return this.errorData.method || null;
    }
}

/**
 * Exception raised for API errors.
 */
class APIError extends KYBError {
    constructor(message, errorData = {}, statusCode = null) {
        super(message, errorData);
        this.statusCode = statusCode;
    }

    /**
     * Get validation errors if any.
     * @returns {Array} Array of validation errors
     */
    getValidationErrors() {
        return this.errorData.error?.validation || [];
    }
}

/**
 * Exception raised for validation errors.
 */
class ValidationError extends APIError {
    constructor(message, errorData = {}) {
        super(message, errorData, 400);
    }

    /**
     * Get detailed validation errors.
     * @returns {Array} Array of validation errors
     */
    getValidationErrors() {
        return this.errorData.error?.validation || [];
    }
}

/**
 * Exception raised for authentication errors.
 */
class AuthenticationError extends APIError {
    constructor(message, errorData = {}) {
        super(message, errorData, 401);
    }
}

/**
 * Exception raised for authorization errors.
 */
class AuthorizationError extends APIError {
    constructor(message, errorData = {}) {
        super(message, errorData, 403);
    }
}

/**
 * Exception raised when a resource is not found.
 */
class NotFoundError extends APIError {
    constructor(message, errorData = {}) {
        super(message, errorData, 404);
    }
}

/**
 * Exception raised when rate limit is exceeded.
 */
class RateLimitError extends APIError {
    constructor(message, errorData = {}) {
        super(message, errorData, 429);
    }
}

/**
 * Exception raised when the service is unavailable.
 */
class ServiceUnavailableError extends APIError {
    constructor(message, errorData = {}) {
        super(message, errorData, 503);
    }
}

/**
 * Exception raised when a request times out.
 */
class TimeoutError extends APIError {
    constructor(message, errorData = {}) {
        super(message, errorData, 408);
    }
}

/**
 * Exception raised for internal server errors.
 */
class InternalError extends APIError {
    constructor(message, errorData = {}) {
        super(message, errorData, 500);
    }
}

module.exports = {
    KYBError,
    APIError,
    ValidationError,
    AuthenticationError,
    AuthorizationError,
    NotFoundError,
    RateLimitError,
    ServiceUnavailableError,
    TimeoutError,
    InternalError
};

File: services/risk-assessment-service/sdks/python/kyb_sdk/exceptions.py
"""
KYB Platform Risk Assessment Service Python SDK Exceptions

This module defines custom exceptions for the KYB Platform SDK.
"""

from typing import Dict, Any, Optional


class KYBError(Exception):
    """Base exception for all KYB SDK errors."""
    
    def __init__(self, message: str, error_data: Optional[Dict[str, Any]] = None):
        super().__init__(message)
        self.message = message
        self.error_data = error_data or {}
    
    def get_request_id(self) -> Optional[str]:
        """Get the request ID from the error data."""
        return self.error_data.get("request_id")
    
    def get_timestamp(self) -> Optional[str]:
        """Get the timestamp from the error data."""
        return self.error_data.get("timestamp")
    
    def get_path(self) -> Optional[str]:
        """Get the request path from the error data."""
        return self.error_data.get("path")
    
    def get_method(self) -> Optional[str]:
        """Get the request method from the error data."""
        return self.error_data.get("method")


class APIError(KYBError):
    """Exception raised for API errors."""
    
    def __init__(
        self,
        message: str,
        error_data: Optional[Dict[str, Any]] = None,
        status_code: Optional[int] = None
    ):
        super().__init__(message, error_data)
        self.status_code = status_code
    
    def get_validation_errors(self) -> list:
        """Get validation errors if any."""
        return self.error_data.get("error", {}).get("validation", [])


class ValidationError(APIError):
    """Exception raised for validation errors."""
    
    def __init__(self, message: str, error_data: Optional[Dict[str, Any]] = None):
        super().__init__(message, error_data, 400)
    
    def get_validation_errors(self) -> list:
        """Get detailed validation errors."""
        return self.error_data.get("error", {}).get("validation", [])


class AuthenticationError(APIError):
    """Exception raised for authentication errors."""
    
    def __init__(self, message: str, error_data: Optional[Dict[str, Any]] = None):
        super().__init__(message, error_data, 401)


class AuthorizationError(APIError):
    """Exception raised for authorization errors."""
    
    def __init__(self, message: str, error_data: Optional[Dict[str, Any]] = None):
        super().__init__(message, error_data, 403)


class NotFoundError(APIError):
    """Exception raised when a resource is not found."""
    
    def __init__(self, message: str, error_data: Optional[Dict[str, Any]] = None):
        super().__init__(message, error_data, 404)


class RateLimitError(APIError):
    """Exception raised when rate limit is exceeded."""
    
    def __init__(self, message: str, error_data: Optional[Dict[str, Any]] = None):
        super().__init__(message, error_data, 429)


class ServiceUnavailableError(APIError):
    """Exception raised when the service is unavailable."""
    
    def __init__(self, message: str, error_data: Optional[Dict[str, Any]] = None):
        super().__init__(message, error_data, 503)


class TimeoutError(APIError):
    """Exception raised when a request times out."""
    
    def __init__(self, message: str, error_data: Optional[Dict[str, Any]] = None):
        super().__init__(message, error_data, 408)


class InternalError(APIError):
    """Exception raised for internal server errors."""
    
    def __init__(self, message: str, error_data: Optional[Dict[str, Any]] = None):
        super().__init__(message, error_data, 500)
